onclick

hXtreme opened the pull request:
- Add UI to enter OnClickURI.
- Update db to store onClickURI and integrate with UI
- Add open URI intent to notifications.

Closes #16

hXtreme committed:
Add UI to enter OnClickURI.

Adds a text field to enter the OnClickURI when creating or editing cron jobs.

hXtreme committed:
Update db to store onClickURI and integrate with UI

- Add onClickURI to db model
- Increment db version to 4
- Add db migration
- UI correctly stores, retrieves, and displays onClickURI.

hXtreme committed:
Add open URI intent to notifications.

Tap notification to open the corresponding URI and auto-dismiss it.

hXtreme commented:
When I was doing this, I noticed that there's a few places that could use some linting and refactoring/renaming work. And there's no other active pr so if you're interested I could do that as well once this is merged.

hXtreme committed:
Remove unnecessary let block.

fi3te commented on app/src/main/java/com/github/fi3te/notificationcron/data/local/NotificationCronInput.kt:
```diff
@@ -0,0 +1,8 @@
+package com.github.fi3te.notificationcron.data.local
+
+data class NotificationCronInput (
```
Reformat file (indentation)

hXtreme replied:
The double indent is actually what I get on reformatting. 8 spaces for continuation indent seem to be the default settings in Android studio.

hXtreme replied:
Fixed.

fi3te commented on app/src/main/java/com/github/fi3te/notificationcron/data/local/NotificationCronInput.kt:
```diff
@@ -0,0 +1,8 @@
+        val cron: String,
+        val notificationTitle: String,
+        val notificationText: String,
+        val onClickURI: String
```
For a better readability I would prefer onClickUri

hXtreme replied:
Done!

fi3te commented on app/src/main/java/com/github/fi3te/notificationcron/ui/NotificationCronAdapter.kt:
```diff
@@ -18,6 +18,7 @@ class NotificationCronAdapter(private var data: List<NotificationCron>, private
         val cronText: TextView = view.findViewById(R.id.cronText)
         val notificationTitleText: TextView = view.findViewById(R.id.notificationTitleText)
         val notificationTextText: TextView = view.findViewById(R.id.notificationTextText)
+        val onClickURI: TextView = view.findViewById(R.id.onClickURIText)
```
id = variable name = onClickUriText

hXtreme replied:
Fixed!

fi3te commented on app/src/main/java/com/github/fi3te/notificationcron/ui/NotificationCronDialog.kt:
```diff
@@ -20,11 +21,12 @@ fun showCreateDialog(windowContext: Context, create: (notificationCron: Notifica
 
         positiveButton(R.string.create) {
-            val (cron, notificationTitle, notificationText) = getInput(this)
+            val (cron, notificationTitle, notificationText, onClickURI) = getInput(this)
```
For a better readability I would prefer onClickUri

hXtreme replied:
Do you think it would be better if instead of unpacking the data class I switch to using with block.

Something like:

```
with (getInput(this)) {
    val notificationCron = NotificatioCron(
        ...
    )
    create(notificationCron)
}
```

hXtreme replied:
For now I've made the change but do let me know what you think about using with block.

fi3te replied:
I have no preference in this respect (too little experience with the Kotlin programming language)

fi3te replied:
If you think that refactoring work is necessary and you are interested, you are welcome to create a PR 😄

fi3te commented on app/src/main/java/com/github/fi3te/notificationcron/ui/NotificationCronDialog.kt:
```diff
@@ -45,11 +47,12 @@ fun showUpdateDialog(
 
         positiveButton(R.string.update) {
-            val (cron, notificationTitle, notificationText) = getInput(this)
+            val (cron, notificationTitle, notificationText, onClickURI) = getInput(this)
```
For a better readability I would prefer onClickUri

hXtreme replied:
Similar to https://github.com/fi3te/NotificationCron/pull/19#discussion_r532101127

fi3te commented on app/src/main/java/com/github/fi3te/notificationcron/ui/NotificationCronDialog.kt:
```diff
@@ -76,15 +79,17 @@ private fun addCronInputView(dialog: MaterialDialog) {
     val notificationTitleInput = customView.findViewById<EditText>(R.id.notificationTitleInput)
     val notificationTextInput = customView.findViewById<EditText>(R.id.notificationTextInput)
-    return Triple(
+    val notificationOnClickURIInput = customView.findViewById<EditText>(R.id.onClickURIInput)
```
id = variable name = onClickUriInput

hXtreme replied:
Fixed!

fi3te commented on app/src/main/res/layout/notification_cron_item.xml:
```diff
@@ -40,6 +40,16 @@
         android:layout_toStartOf="@+id/notificationCronOptionsButton"
         android:textAppearance="@style/ListItemSecondaryTextAppearance" />
 
+    <TextView
```
The marginBottom of the above element must be removed

hXtreme replied:
Done.

fi3te commented on app/src/main/res/values/strings.xml:
```diff
@@ -6,6 +6,7 @@
     <string name="cron_input_hint">Cron expression (Quartz)</string>
     <string name="notification_title_input_hint">Notification title</string>
     <string name="notification_text_input_hint">Notification text</string>
+    <string name="on_click_uri_input_hint">On click open uri</string>
```
I would prefer "URI (opens on click)", "URI (öffnet sich bei Klick)", "URI (opent op klik)"
The German and Dutch translation is missing

hXtreme replied:
I've made the changes and added the translation, though you should probably give it a look as I don't speak either of those languages.

fi3te reviewed (changes requested):
Thank you for the pull request 😄

fi3te commented on app/src/main/java/com/github/fi3te/notificationcron/ui/NotificationCronAdapter.kt:
```diff
@@ -68,6 +69,7 @@ class NotificationCronAdapter(private var data: List<NotificationCron>, private
             cronText.text = notificationCron.cron
             notificationTitleText.text = notificationCron.notificationTitle
             notificationTextText.text = notificationCron.notificationText
+            onClickURI.text = notificationCron.onClickURI
```
Maybe this field should be hidden if it is not used (View.GONE)

hXtreme replied:
I tried it out initially but felt that having different heights for items in the list was a bit off.

On top of that the same isn't being done for the Notification text field.

fi3te replied:
Very long notification texts are possible, which can lead to different heights. But you are right, using the fields differently would also be inconsistent.

hXtreme committed:
Revert changes made to 3.json

hXtreme committed:
Changes based on code review.

- Rename onClickURI to onClickUri.
- Fix variable names.
- Update on_click_uri_input_hint based on review and add suggested
translations.
- Change indentation of NotificationCronInput to fit the poject style.
- Remove marginBottom from notificationTextText.

hXtreme commented on app/schemas/com.github.fi3te.notificationcron.data.local.AppDatabase/3.json:
```diff
@@ -2,11 +2,11 @@
   "formatVersion": 1,
```
Changes reverted.

hXtreme commented on app/schemas/com.github.fi3te.notificationcron.data.local.AppDatabase/4.json:
```diff
@@ -0,0 +1,76 @@
+            "notNull": true
+          },
+          {
+            "fieldPath": "onClickURI",
```
Done!

fi3te reviewed (approved):
Thank you very much!
